### PR TITLE
Fix System.IO.Pipes bugs (src and test)

### LINKED ISSUE #1762: Unix: System.IO.Pipes; Operations not throwing exceptions after disconnect
```C#
 [Fact]
        public static void ServerAfterDisconnectThrows()
        {
            using (NamedPipeServerStream server = new NamedPipeServerStream("unique3", PipeDirection.InOut, 1, PipeTransmissionMode.Byte, PipeOptions.Asynchronous))
            {
                using (NamedPipeClientStream client = new NamedPipeClientStream("unique3"))
                {
                    Task clientConnect = client.ConnectAsync(); 
                    server.WaitForConnection();
                    Assert.Throws<InvalidOperationException>(() => server.IsMessageComplete);
                    Assert.Throws<InvalidOperationException>(() => server.WaitForConnection());
                    Assert.ThrowsAsync<InvalidOperationException>(() => server.WaitForConnectionAsync());

                    server.Disconnect();
                    Assert.Throws<InvalidOperationException>(() => server.Disconnect());    // double disconnect
                    Assert.Throws<InvalidOperationException>(() => server.WriteByte(5));
                    Assert.Throws<InvalidOperationException>(() => server.ReadByte());
                    Assert.Throws<InvalidOperationException>(() => server.IsMessageComplete);
                }
            }
        }
```

13:30:02 �[39;49m�[31m   System.IO.Pipes.Tests.NamedPipesThrowsTests.ServerAfterDisconnectThrows [FAIL]
13:30:02 �[37m      Assert.Throws() Failure
13:30:02       Expected: typeof(System.InvalidOperationException)
13:30:02       Actual:   (No exception was thrown)
13:30:02 �[30m      Stack Trace:
13:30:02 �[37m            at System.IO.Pipes.Tests.NamedPipesThrowsTests.ServerAfterDisconnectThrows()
13:30:03 �[31m   System.IO.Pipes.Tests.NamedPipesThrowsTests.ServerWhenDisplosedThrows [FAIL]
13:30:03 �[37m      Assert.Throws() Failure
13:30:03       Expected: typeof(System.ObjectDisposedException)
13:30:03       Actual:   typeof(System.InvalidOperationException): Pipe handle has not been set.  Did your PipeStream implementation call InitializeHandle?
13:30:03 �[30m      Stack Trace:
13:30:03 �[37m            at System.IO.Pipes.NamedPipeServerStream.CheckDisconnectOperations()
13:30:03             at System.IO.Pipes.NamedPipeServerStream.Disconnect()
13:30:03             at System.IO.Pipes.Tests.NamedPipesThrowsTests.<>c__DisplayClass64.<ServerWhenDisplosedThrows>b__5a()

### PR DESCRIPTION
**Pull request by stephentoub**
Fix multiple test bugs in System.IO.Pipes.Tests and multiple product bugs in the Unix implementation of System.IO.Pipes, primarily around error cases surfacing an exception or the right exception and at the right time.

Product changes:
- In the Windows implementation, avoids a few unnecessary allocations when registring for cancellation.
- Common code shared between Windows and Unix is used to check the state of the streams prior to various operations, including validating the current state of the pipe's handle.  On Windows, the handle is always initialized at creation, whereas in the current Unix implementation, the handle's creation is deferred until the connection is established.  As such, code that checked the validity of the handle needed to be updated to not require that the handle be set on Unix.
- The Unix NamedPipeServerStream.WaitForConnection implementation was missing a check to fail if the connection was already established.  On Windows, this is handled by the underlying OS function; on Unix, we just open a file, which any number of consumers can do, so there's no valdation provided by the OS for the constraint we need.
- The Unix implementation of PipeStream was not overriding Read/WriteAsync, such that validation of the inputs was happening asynchronously.  I've refactored out the Read/WriteAsync validation logic to be shared between Windows and Unix, with the shared code overriding these methods.
- In the Unix implementation, if any operation detects that the pipe has been broken, it needs to transition the pipe to a state of Broken.  The Windows implementation was already doing this.

Test changes:
- A bunch of calls to Assert.ThrowsAsync were ignoring the returned Task, which meant that a) any assertion failures were being ignored, b) operations may have continued to run well past the end of the test, and c) race conditions were occurring due to unexpected concurrency between operations.  These have been addressed by either awaiting on the resulting task (when the operation is expected to run asynchronously) or by changing to using Throw instead of ThrowAsync (when we expect the exception to propagate out synchronously).
- Some tasks were being waited on synchronously, which could have potentially led to deadlock situations in the tests.  I changed everything relevant to use async/await.
- The ServerCloneTests was failing (on both Windows and Unix) because it was closing the client stream prior to being done with the test.  I fixed up the disposal ordering and re-enabled the test. I also had to change the order in which operations occurred around waiting for the connection, due to the Unix implementation not establishing the handle until the connection is established, rather than during construction.
- I added some Interop.IsWindows checks for functionality known to be unavailable on Unix, e.g. NumberOfServerInstances.
- I changed a few tests to use In vs Out rather than InOut, to focus the test on what specifically was being tested, and because the Unix support for InOut isn't quite the same as on Windows.

Fixes #1762, #1763, #1764, #1765, #1766, #1772.

(This commit also currently includes the changes in #1871; I will fix those appropriately once #1871 is merged.)

### CODE REVIEW

**Comment by bpschoch** [comment]
wow great work.  I have a version almost ready to go that refactors many of the throws that should be done on each pipe variant into a common base class.  I'll wait until this commits and merge your changes in.

Also when I did the tests, I wondered why on the xxxDisconnectedPipeThrows test, after a disconnect, the write type operations throw exceptions while the read type don't.  It seemed strange that they both wouldn't do the same.

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by stephentoub** [comment]
@dotnet-bot test this please

File: src/System.IO.Pipes/tests/AnonymousPipesThrowsTests.cs [hunk 1]
```diff
@@ -1,17 +1,15 @@
...
-using Xunit;
-
 using Microsoft.Win32.SafeHandles;
+using Xunit;
 
 namespace System.IO.Pipes.Tests
 {
     public class AnonymousPipesThrowsTests
     {
+        private static void NotReachable(object obj) { Assert.True(false, "This should not be reached."); }
```
Comment (KrzysztofCwalina) [inline src/System.IO.Pipes/tests/AnonymousPipesThrowsTests.cs]: Any reason you are not using Assert.Fail here?

File: src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs [hunk 2]
```diff
@@ -62,7 +71,7 @@ internal static string GetPipePath(string serverName, string pipeName)
 
         /// <summary>Throws an exception if the supplied handle does not represent a valid pipe.</summary>
         /// <param name="safePipeHandle">The handle to validate.</param>
-        internal static void ValidateHandleIsPipe(SafePipeHandle safePipeHandle)
+        internal void ValidateHandleIsPipe(SafePipeHandle safePipeHandle)
```
Comment (KrzysztofCwalina) [inline src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs]: Why did you change it to an instance method?

File: src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Windows.cs [hunk 3]
```diff
@@ -120,6 +128,13 @@ private unsafe void WriteCore(byte[] buffer, int offset, int count)
...
+        [SecuritySafeCritical]
+        private Task WriteAsyncCore(Byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            ReadWriteAsyncParams state = new ReadWriteAsyncParams(buffer, offset, count, cancellationToken);
+            return Task.Factory.FromAsync(BeginWrite, EndWrite, state);
+        }
+
         // Blocks until the other end of the pipe has read in all written buffer.
         [SecurityCritical]
         public void WaitForPipeDrain()
```
Comment (KrzysztofCwalina) [inline src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Windows.cs]: First time I read it, I misread it as "WaitForPipeDream" :-)

File: src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs [hunk 4]
```diff
@@ -292,9 +327,23 @@ internal static Interop.libc.OpenFlags TranslateFlags(PipeDirection direction, P
                 int fd = (int)handle.DangerousGetHandle();
                 Debug.Assert(fd >= 0);
 
-                long result;
-                while (Interop.CheckIo(result = sysCall(fd, arg1, arg2))) ;
-                return result;
+                while (true)
```
Comment (KrzysztofCwalina) [inline src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs]: Should we worry about this silently never terminating?

File: src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs (hunk 2 above)
Comment (stephentoub) [inline src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs]: Because SysCall that it calls became an instance method, too.

File: src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs (hunk 4 above)
Comment (stephentoub) [inline src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs]: I don't think so. The only way that'd happen is if every system call got interrupted each and every time.

File: src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Windows.cs (hunk 3 above)
Comment (stephentoub) [inline src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Windows.cs]: Heh :smile:

File: src/System.IO.Pipes/tests/AnonymousPipesThrowsTests.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.IO.Pipes/tests/AnonymousPipesThrowsTests.cs]: Because there's no such method? :smile:

Comment (bpschoch) [inline src/System.IO.Pipes/tests/AnonymousPipesThrowsTests.cs]: Perhaps XUnit needs a call like Assert.ThrowsAsyncImmediate()  To ensure the throw happens before the task is created (I assume that is the purpose of NotReachable())

Comment (stephentoub) [inline src/System.IO.Pipes/tests/AnonymousPipesThrowsTests.cs]: > I assume that is the purpose of NotReachable()

Yes.  Assert.Throws would actually be the right thing here, except that as a "help" it detects that the object returned is a Task and throws an exception to say that Task-returning methods should instead use Async.ThrowsAsync.

**Comment by ericstj** [comment]
I reviewed the product code and skimmed the test code and it looks ok to me.  :+1: 
/cc @krwq who did the initial test port.

**Comment by stephentoub** [comment]
Thanks Eric and Krzysztof C. for reviewing.  Other Krzysztof, any additional thoughts?

File: src/System.IO.Pipes/tests/AnonymousPipesThrowsTests.cs (hunk 1 above)
Comment (KrzysztofCwalina) [inline src/System.IO.Pipes/tests/AnonymousPipesThrowsTests.cs]: Oops. I confused XUnit with VS test.

Comment (stephentoub) [inline src/System.IO.Pipes/tests/AnonymousPipesThrowsTests.cs]: :smile:

File: src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs (hunk 2 above)
Comment (KrzysztofCwalina) [inline src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs]: And why did SysCall become an instance method?

Comment (ericstj) [inline src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs]: Because he's updating the (instance) pipe state when he gets an error.

Comment (stephentoub) [inline src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs]: Right. Is there some reason I've missed that this shouldn't be an instance method?

Comment (KrzysztofCwalina) [inline src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.Unix.cs]: Ah, I see it now. Thanks.

File: src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.cs [hunk 5]
```diff
@@ -166,6 +196,35 @@ public override void Write(byte[] buffer, int offset, int count)
             return;
         }
 
+        [SecuritySafeCritical]
+        public override Task WriteAsync(Byte[] buffer, int offset, int count, CancellationToken cancellationToken)
```
Comment (krwq) [inline src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.cs]: nit: byte[] (low case)

**Comment by krwq** [comment]
LGTM :+1:

File: src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.cs (hunk 5 above)
Comment (stephentoub) [inline src/System.IO.Pipes/src/System/IO/Pipes/PipeStream.cs]: Thanks, fixed.

**Comment by stephentoub** [comment]
> wow great work. I have a version almost ready to go that refactors many of the throws that should be done on each pipe variant into a common base class. I'll wait until this commits and merge your changes in.

Thanks :smile:
